### DISCUSSION TITLE
macOS12 can crash

### DISCUSSION THREAD

**Discussion by sl1200mk2** [libuv/libuv#4661] [Q&A]
* **Version**: latest
* **Platform**: macOS 12.7
Hi,
here's more or less my code (in a for loop):
send_request_t* send_req = NULL;
if(NULL != (this_Client = loop_ptr->uv_myWriters[i]) &&
               NULL != this_Client->ch &&
               NULL != (send_req = (send_request_t*) malloc(sizeof(send_request_t)))) {
                send_req->buf = uv_buf_init((char*) malloc(length), (unsigned int)length);
                if (send_req->buf.base) {
                    memcpy(send_req->buf.base, msg, length);
                    uv_udp_send(&send_req->req, this_Client->ch, &send_req->buf, 1, &this_Client->cliAddr.sa, afterSending_sRequest);
                } else {
                    free(send_req);
                }
}

this works fine on my macOS15.2(Mx) but on macOS12 (Intel), a user is reporting a crash like that:
System Integrity Protection: enabled

Crashed Thread:        9

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000008
Exception Codes:       0x0000000000000001, 0x0000000000000008
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [5796]

VM Region Info: 0x8 is not in any region.  Bytes before following region: 140737488289784
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      VM_ALLOCATE              7fffffff0000-7fffffff1000 [    4K] r-x/r-x SM=ALI  

....

Thread 9 Crashed:
0   libuv.1.0.0.dylib             	       0x1026812f7 uv__udp_sendmsg + 407
1   libuv.1.0.0.dylib             	       0x102681015 uv__udp_send + 341
2   libuv.1.0.0.dylib             	       0x10267032e uv_udp_send + 126

it's not every time, it happens randomly.
if I change my code to use  :

           if(NULL != (this_Client = loop_ptr->uv_myWriters[i]) &&
               NULL != this_Client->ch) {
                uv_buf_t buf = uv_buf_init(msg, (unsigned int)length);
                uv_udp_try_send(this_Client->ch, &buf, 1, &this_Client->cliAddr.sa);
             }

there's no crash....

does it means something to you?
is there's a way to detail the crashLog?

best regards,
nicolas

**Comment by saghul** [comment]
Are you calling that code from a thread different than the one running the loop?

**Comment by sl1200mk2** [comment]
Hi,
nope never... I post the request thanks to uv_async_send(&x->uvLoop_ptr->uvSyncT_ExtEvents_runner);
then in the runner I use uv_queue_work()

++

**Comment by saghul** [comment]
Can you post a full example? You cannot call loop APIs from the work callback because that runs in another thread.

If you are instead calling async_send from the work callback, take into account that one call to send need not result in a call to the callback.

**Comment by sl1200mk2** [comment]
thank you for your answers...
the full example is a little bit complicated.

first I register the async callback:
uv_async_init(loop_ptr->server_loop, &(loop_ptr->uvSyncT_ExtEvents_runner), syncUV_ExtEvents_runner_cb);

then later I 
uv_async_send(&x->uvLoop_ptr->uvSyncT_ExtEvents_runner);

the callback is :
void syncUV_ExtEvents_runner_cb(uv_async_t *handle) {
    t_uv_loopMag* loop_ptr = NULL;
    t_ptr* x = NULL;
    struct asyncWork_baton *baton = NULL;

    if(NULL != (loop_ptr = handle->data) &&
          NULL != (x = (t_ptr*) loop_ptr->t_dl_ptr)) {

             if(NULL != (baton = uv_baton_new())) {
                baton->dl_ptr = x;
                baton->req.data = baton;
                baton->bdh_ptr = NULL;
                baton->l_ptr = NULL;
                baton->message_fPtr = x->externalMessagingSystem;
                uv_queue_work(loop_ptr->server_loop, &baton->req, syncUV_SendingExternalPendingEvents, syncUV_QueueWork_done);
            }
    }
    return;
}

and after in the static void syncUV_SendingExternalPendingEvents(uv_work_t *req); there's the for loop which randomly crash on macOS

**Comment by sl1200mk2** [comment]
mmm, you mean that:
if(NULL != (this_Client = loop_ptr->uv_myWriters[i]) &&
           NULL != this_Client->ch) {
            uv_buf_t buf = uv_buf_init(msg, (unsigned int)length);
            uv_udp_try_send(this_Client->ch, &buf, 1, &this_Client->cliAddr.sa);
         }

should be used in syncUV_QueueWork_done() and not in syncUV_SendingExternalPendingEvents(), isn't it?

**Comment by sl1200mk2** [comment]
you were right !!
thank you for advices.

++
